openvdb/points/PointScatter.h

```diff
@@ -0,0 +1,676 @@
+    public BasePointScatter<RandGenT, ValueType, CodecT, PointDataGridT, InterruptT>
+{
+public:
+    using BaseT = BasePointScatter<RandGenT, ValueType, CodecT, PointDataGridT, InterruptT>;
+
+    UniformVoxelPointScatter(const float pointsPerVoxel,
+                             const Index64 seed,
+                             const float spread = 1.0f,
+                             InterruptT* interrupter = nullptr)
+        : BaseT(seed, spread, interrupter)
+        , mPointsPerVoxel(math::Floor(pointsPerVoxel))
+        , mDelta(pointsPerVoxel - float(mPointsPerVoxel))
+        , mFractional(!math::isApproxZero(mDelta, 1.0e-6)) {
+            if (pointsPerVoxel < 0.0f) {
+                OPENVDB_THROW(ValueError, "Points per voxel must not be less than zero.");
+            }
+        }
+
+    /// @brief Execute uniform voxel point scattering for a given grid of any type
+    /// @note only the grids topology is used
+    /// @parm grid   The source grid to base the topology generation from
+    template<typename GridT>
+    void operator()(const GridT& grid)
+    {
+        BaseT::start("Dense uniform scattering with fixed point count");
+        if (!BaseT::init(grid)) return;
+        const bool fractionalOnly(mPointsPerVoxel == 0);
+        if (fractionalOnly && !mFractional) {
+            BaseT::mPoints->clear();
+            return;
+        }
+
+        typename BaseT::LeafManagerT leafManager(BaseT::mPoints->tree());
+
+        leafManager.foreach([&](typename BaseT::LeafNodeT& leaf, const size_t) {
+            if (BaseT::wasInterrupted()) return;
+            Index32 offset(0);
+            for (auto iter = leaf.beginValueAll(); iter; ++iter) {
+                if (iter.isValueOn()) {
+                    offset += mPointsPerVoxel;
+                    if (mFractional && BaseT::getRand01() < mDelta) ++offset;
+                    else if (fractionalOnly) leaf.setValueOff(iter.pos());
+                }
+                // @note can't use iter.setValue(offset) on point grids
+                leaf.setOffsetOnly(iter.pos(), offset);
+            }
+
+            if (offset != 0) BaseT::generatePositions(leaf, offset);
+        }, /*threaded*/false);
+
+        if (fractionalOnly) tools::pruneInactive(BaseT::mPoints->tree());
+        BaseT::end();
+    }
+
+    /// @brief Print information about the scattered points
+    /// @parm name  A name to insert into the printed info
+    /// @parm os    The output stream
+    void print(const std::string &name, std::ostream& os = std::cout) const
+    {
+        if (!BaseT::mPoints) return;
+        os << "Uniformly scattered " << pointCount(BaseT::mPoints->tree())
+           << " points into " << BaseT::mPoints->activeVoxelCount()
+           << " active voxels in \"" << name << "\" corresponding to "
+           << mPointsPerVoxel << " points per voxel." << std::endl;
+    }
+
+private:
+    const Index32 mPointsPerVoxel;
+    const double mDelta;
+    const bool mFractional;
+};
```

Comment:
Not a big deal, but I find it quite useful when navigating the code to have comments such as "}; // class UniformVoxelPointScatter".